File: backend/routers/card_trends.py
"""
card_trends.py — Card portfolio analytics
Endpoints for the Card Portfolio Report: issuance trend, active/inactive
portfolio health, card program (BLINK) breakdown, status distribution.
"""
import re
from fastapi import APIRouter, Depends, Query, HTTPException
from typing import Optional
from core.database import get_db_pg, get_db_mssql
from core.auth import require_pages
from core.dual_query import dual_query, dual_scalar

router = APIRouter()
ACCESS = require_pages(["card_trends"])

_DATE_RE = re.compile(r'^\d{4}-\d{2}-\d{2}$')

def _vd(d, name):
    if d is None: return None
    if not _DATE_RE.match(d): raise HTTPException(400, f"Invalid {name} — YYYY-MM-DD required")
    return d

# Date filter on Account_Created_Date
def _df_ms(df, dt):
    if df and dt: return f" AND CAST(Account_Created_Date AS DATE) BETWEEN '{df}' AND '{dt}'"
    if df:        return f" AND CAST(Account_Created_Date AS DATE) >= '{df}'"
    if dt:        return f" AND CAST(Account_Created_Date AS DATE) <= '{dt}'"
    return ""

def _df_pg(df, dt):
    if df and dt: return f' AND "Account Created Date"::date BETWEEN \'{df}\' AND \'{dt}\''
    if df:        return f' AND "Account Created Date"::date >= \'{df}\''
    if dt:        return f' AND "Account Created Date"::date <= \'{dt}\''
    return ""

def _filters(product, card_program, df, dt):
    ms = _df_ms(df, dt)
    pg = _df_pg(df, dt)
    params: dict = {}
    if product:
        ms += " AND Product_Name=:product"
        pg += ' AND "Product Name"=:product'
        params["product"] = product
    if card_program:
        ms += " AND Card_Product=:card_program"
        pg += ' AND "Card Product"=:card_program'
        params["card_program"] = card_program
    return ms, pg, params


# ── Portfolio KPIs ────────────────────────────────────────────────────────────

@router.get("/kpis")
def kpis(
    product:      Optional[str] = Query(None),
    card_program: Optional[str] = Query(None),
    date_from:    Optional[str] = Query(None),
    date_to:      Optional[str] = Query(None),
    db_pg=Depends(get_db_pg), db_mssql=Depends(get_db_mssql), user=Depends(ACCESS)
):
    date_from = _vd(date_from, "date_from"); date_to = _vd(date_to, "date_to")
    fms, fpg, fp = _filters(product, card_program, date_from, date_to)
    k, s = {}, []

    def q(ms, pg, key, cast=int):
        val, src = dual_scalar(db_mssql, db_pg, ms, pg, params=fp)
        k[key] = cast(val) if val is not None else 0
        s.append(src)

    q(f"SELECT COUNT(*) AS val FROM dbo.Account WHERE 1=1{fms}",
      f'SELECT COUNT(*) AS val FROM "Products" WHERE 1=1{fpg}', "total_issued")

    q(f"SELECT COUNT(*) AS val FROM dbo.Account WHERE Status IN ('Open','Active'){fms}",
      f'SELECT COUNT(*) AS val FROM "Products" WHERE "Account Status" IN (\'Open\',\'Active\'){fpg}', "active")

    q(f"SELECT COUNT(*) AS val FROM dbo.Account WHERE Status NOT IN ('Open','Active'){fms}",
      f'SELECT COUNT(*) AS val FROM "Products" WHERE "Account Status" NOT IN (\'Open\',\'Active\'){fpg}', "inactive")

    q(f"SELECT COUNT(*) AS val FROM dbo.Account WHERE Status='TERMINATED'{fms}",
      f'SELECT COUNT(*) AS val FROM "Products" WHERE "Account Status"=\'TERMINATED\'{fpg}', "terminated")

    q(f"SELECT COUNT(*) AS val FROM dbo.Account WHERE Status IN ('LEGAL ACTI','SUSPENDED'){fms}",
      f'SELECT COUNT(*) AS val FROM "Products" WHERE "Account Status" IN (\'LEGAL ACTI\',\'SUSPENDED\'){fpg}', "legal_suspended")

    # MTD always uses current month regardless of date filter
    q(f"SELECT COUNT(*) AS val FROM dbo.Account WHERE MONTH(Account_Created_Date)=MONTH(GETDATE()) AND YEAR(Account_Created_Date)=YEAR(GETDATE())",
      f'SELECT COUNT(*) AS val FROM "Products" WHERE DATE_TRUNC(\'month\',"Account Created Date")=DATE_TRUNC(\'month\',CURRENT_DATE)', "issued_mtd")

    k["activation_rate"] = round(k["active"] / k["total_issued"] * 100, 1) if k["total_issued"] > 0 else 0
    return {"data": k, "data_source": "mssql_live" if "mssql_live" in s else "supabase_snapshot"}


# ── Monthly issuance trend ────────────────────────────────────────────────────

@router.get("/issuance-trend")
def issuance_trend(
    product:      Optional[str] = Query(None),
    card_program: Optional[str] = Query(None),
    date_from:    Optional[str] = Query(None),
    date_to:      Optional[str] = Query(None),
    db_pg=Depends(get_db_pg), db_mssql=Depends(get_db_mssql), user=Depends(ACCESS)
):
    date_from = _vd(date_from, "date_from"); date_to = _vd(date_to, "date_to")
    fms, fpg, fp = _filters(product, card_program, date_from, date_to)
    data, src = dual_query(db_mssql, db_pg,
        f"""SELECT FORMAT(Account_Created_Date,'MMM yyyy') AS month,
               DATEFROMPARTS(YEAR(Account_Created_Date),MONTH(Account_Created_Date),1) AS month_sort,
               COUNT(*) AS issued
            FROM dbo.Account WHERE Account_Created_Date IS NOT NULL{fms}
            GROUP BY DATEFROMPARTS(YEAR(Account_Created_Date),MONTH(Account_Created_Date),1),
                     FORMAT(Account_Created_Date,'MMM yyyy')
            ORDER BY month_sort""",
        f"""SELECT TO_CHAR(DATE_TRUNC('month',"Account Created Date"),'Mon YYYY') AS month,
               DATE_TRUNC('month',"Account Created Date") AS month_sort,
               COUNT(*) AS issued
            FROM "Products" WHERE "Account Created Date" IS NOT NULL{fpg}
            GROUP BY DATE_TRUNC('month',"Account Created Date") ORDER BY month_sort""",
        params=fp)
    return {"data": data, "data_source": src}


# ── Active/Inactive portfolio health by issuance cohort ──────────────────────

@router.get("/portfolio-health")
def portfolio_health(
    product:      Optional[str] = Query(None),
    card_program: Optional[str] = Query(None),
    date_from:    Optional[str] = Query(None),
    date_to:      Optional[str] = Query(None),
    db_pg=Depends(get_db_pg), db_mssql=Depends(get_db_mssql), user=Depends(ACCESS)
):
    """Monthly cohort view: of cards issued each month, how many are active vs inactive today."""
    date_from = _vd(date_from, "date_from"); date_to = _vd(date_to, "date_to")
    fms, fpg, fp = _filters(product, card_program, date_from, date_to)
    data, src = dual_query(db_mssql, db_pg,
        f"""SELECT FORMAT(Account_Created_Date,'MMM yyyy') AS month,
               DATEFROMPARTS(YEAR(Account_Created_Date),MONTH(Account_Created_Date),1) AS month_sort,
               SUM(CASE WHEN Status IN ('Open','Active') THEN 1 ELSE 0 END) AS active,
               SUM(CASE WHEN Status NOT IN ('Open','Active') THEN 1 ELSE 0 END) AS inactive,
               COUNT(*) AS total
            FROM dbo.Account WHERE Account_Created_Date IS NOT NULL{fms}
            GROUP BY DATEFROMPARTS(YEAR(Account_Created_Date),MONTH(Account_Created_Date),1),
                     FORMAT(Account_Created_Date,'MMM yyyy')
            ORDER BY month_sort""",
        f"""SELECT TO_CHAR(DATE_TRUNC('month',"Account Created Date"),'Mon YYYY') AS month,
               DATE_TRUNC('month',"Account Created Date") AS month_sort,
               SUM(CASE WHEN "Account Status" IN ('Open','Active') THEN 1 ELSE 0 END) AS active,
               SUM(CASE WHEN "Account Status" NOT IN ('Open','Active') THEN 1 ELSE 0 END) AS inactive,
               COUNT(*) AS total
            FROM "Products" WHERE "Account Created Date" IS NOT NULL{fpg}
            GROUP BY DATE_TRUNC('month',"Account Created Date") ORDER BY month_sort""",
        params=fp)
    return {"data": data, "data_source": src}


# ── Status distribution (donut data) ─────────────────────────────────────────

@router.get("/status-distribution")
def status_distribution(
    product:      Optional[str] = Query(None),
    card_program: Optional[str] = Query(None),
    date_from:    Optional[str] = Query(None),
    date_to:      Optional[str] = Query(None),
    db_pg=Depends(get_db_pg), db_mssql=Depends(get_db_mssql), user=Depends(ACCESS)
):
    date_from = _vd(date_from, "date_from"); date_to = _vd(date_to, "date_to")
    fms, fpg, fp = _filters(product, card_program, date_from, date_to)
    data, src = dual_query(db_mssql, db_pg,
        f"SELECT Status AS status, COUNT(*) AS count FROM dbo.Account WHERE 1=1{fms} GROUP BY Status ORDER BY count DESC",
        f'SELECT "Account Status" AS status, COUNT(*) AS count FROM "Products" WHERE 1=1{fpg} GROUP BY "Account Status" ORDER BY count DESC',
        params=fp)
    return {"data": data, "data_source": src}


# ── Card Program (BLINK) breakdown ────────────────────────────────────────────

@router.get("/by-program")
def by_program(
    date_from: Optional[str] = Query(None),
    date_to:   Optional[str] = Query(None),
    db_pg=Depends(get_db_pg), db_mssql=Depends(get_db_mssql), user=Depends(ACCESS)
):
    date_from = _vd(date_from, "date_from"); date_to = _vd(date_to, "date_to")
    fms, fpg, fp = _filters(None, None, date_from, date_to)
    try:
        data, src = dual_query(db_mssql, db_pg,
            f"""SELECT Card_Product AS program,
                   COUNT(*) AS total,
                   SUM(CASE WHEN Status IN ('Open','Active') THEN 1 ELSE 0 END) AS active,
                   SUM(CASE WHEN Status NOT IN ('Open','Active') THEN 1 ELSE 0 END) AS inactive,
                   ROUND(100.0 * SUM(CASE WHEN Status IN ('Open','Active') THEN 1 ELSE 0 END) / COUNT(*), 1) AS activation_rate
                FROM dbo.Account WHERE Card_Product IS NOT NULL AND Card_Product != ''{fms}
                GROUP BY Card_Product ORDER BY total DESC""",
            f"""SELECT "Card Product" AS program,
                   COUNT(*) AS total,
                   SUM(CASE WHEN "Account Status" IN ('Open','Active') THEN 1 ELSE 0 END) AS active,
                   SUM(CASE WHEN "Account Status" NOT IN ('Open','Active') THEN 1 ELSE 0 END) AS inactive,
                   ROUND(100.0 * SUM(CASE WHEN "Account Status" IN ('Open','Active') THEN 1 ELSE 0 END) / COUNT(*), 1) AS activation_rate
                FROM "Products" WHERE "Card Product" IS NOT NULL AND "Card Product" != ''{fpg}
                GROUP BY "Card Product" ORDER BY total DESC""",
            params=fp)
        return {"data": data, "data_source": src}
    except Exception:
        return {"data": [], "data_source": "supabase_snapshot"}


# ── Product breakdown ─────────────────────────────────────────────────────────

@router.get("/by-product")
def by_product(
    card_program: Optional[str] = Query(None),
    date_from:    Optional[str] = Query(None),
    date_to:      Optional[str] = Query(None),
    db_pg=Depends(get_db_pg), db_mssql=Depends(get_db_mssql), user=Depends(ACCESS)
):
    date_from = _vd(date_from, "date_from"); date_to = _vd(date_to, "date_to")
    fms, fpg, fp = _filters(None, card_program, date_from, date_to)
    data, src = dual_query(db_mssql, db_pg,
        f"""SELECT Product_Name,
               COUNT(*) AS total,
               SUM(CASE WHEN Status IN ('Open','Active') THEN 1 ELSE 0 END) AS active,
               SUM(CASE WHEN Status NOT IN ('Open','Active') THEN 1 ELSE 0 END) AS inactive,
               ROUND(100.0 * SUM(CASE WHEN Status IN ('Open','Active') THEN 1 ELSE 0 END) / COUNT(*), 1) AS activation_rate
            FROM dbo.Account WHERE Product_Name IS NOT NULL{fms}
            GROUP BY Product_Name ORDER BY total DESC""",
        f"""SELECT "Product Name",
               COUNT(*) AS total,
               SUM(CASE WHEN "Account Status" IN ('Open','Active') THEN 1 ELSE 0 END) AS active,
               SUM(CASE WHEN "Account Status" NOT IN ('Open','Active') THEN 1 ELSE 0 END) AS inactive,
               ROUND(100.0 * SUM(CASE WHEN "Account Status" IN ('Open','Active') THEN 1 ELSE 0 END) / COUNT(*), 1) AS activation_rate
            FROM "Products" WHERE "Product Name" IS NOT NULL{fpg}
            GROUP BY "Product Name" ORDER BY total DESC""",
        params=fp)
    return {"data": data, "data_source": src}


# ── List distinct card programs (for filter dropdown) ────────────────────────

@router.get("/programs")
def list_programs(
    db_pg=Depends(get_db_pg), db_mssql=Depends(get_db_mssql), user=Depends(ACCESS)
):
    try:
        data, src = dual_query(db_mssql, db_pg,
            "SELECT DISTINCT Card_Product AS program FROM dbo.Account WHERE Card_Product IS NOT NULL AND Card_Product != '' ORDER BY Card_Product",
            'SELECT DISTINCT "Card Product" AS program FROM "Products" WHERE "Card Product" IS NOT NULL AND "Card Product" != \'\' ORDER BY "Card Product"')
        return {"data": [r["program"] for r in data], "data_source": src}
    except Exception:
        return {"data": [], "data_source": "supabase_snapshot"}

File: backend-go/handlers/card_trends.go
package handlers

import (
	"fmt"
	"net/http"

	"github.com/go-chi/chi/v5"
	"github.com/o3c/reports/core"
)

func RegisterCardTrends(r chi.Router, db *core.DB) {
	r.Use(core.RequirePages("card_trends"))
	r.Get("/kpis", cardTrendsKPIs(db))
	r.Get("/issuance-trend", cardTrendsIssuance(db))
	r.Get("/portfolio-health", cardTrendsPortfolioHealth(db))
	r.Get("/status-distribution", cardTrendsStatusDist(db))
	r.Get("/by-program", cardTrendsByProgram(db))
	r.Get("/by-product", cardTrendsByProduct(db))
	r.Get("/programs", cardTrendsPrograms(db))
}

// buildTrendsFilter builds the WHERE filter for card trend endpoints.
// Account_Created_Date / opened_date is the date column.
func buildTrendsFilter(r *http.Request) (*Filter, error) {
	dateFrom, err := validDate(r, "date_from")
	if err != nil {
		return nil, err
	}
	dateTo, err := validDate(r, "date_to")
	if err != nil {
		return nil, err
	}
	product := qstr(r, "product")
	cardProgram := qstr(r, "card_program")

	var f Filter
	f.Date("Account_Created_Date", `opened_date`, dateFrom, dateTo)
	f.Eq(" AND Product_Name=?", ` AND product_name=?`, product)
	f.Eq(" AND Card_Product=?", ` AND COALESCE(card_product,card_program)=?`, cardProgram)
	return &f, nil
}

func cardTrendsKPIs(db *core.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		f, err := buildTrendsFilter(r)
		if err != nil {
			respondErr(w, 400, err.Error())
			return
		}

		ctx := r.Context()
		kpis := map[string]any{}
		var sources []string

		type spec struct{ key, pg string }
		for _, s := range []spec{
			{"total_issued",
				fmt.Sprintf(`SELECT COUNT(*) AS val FROM app.accounts WHERE 1=1%s`, f.PG())},
			{"active",
				fmt.Sprintf(`SELECT COUNT(*) AS val FROM app.accounts WHERE status IN ('Open','Active')%s`, f.PG())},
			{"inactive",
				fmt.Sprintf(`SELECT COUNT(*) AS val FROM app.accounts WHERE status NOT IN ('Open','Active')%s`, f.PG())},
			{"terminated",
				fmt.Sprintf(`SELECT COUNT(*) AS val FROM app.accounts WHERE status='TERMINATED'%s`, f.PG())},
			{"legal_suspended",
				fmt.Sprintf(`SELECT COUNT(*) AS val FROM app.accounts WHERE status IN ('LEGAL ACTI','SUSPENDED')%s`, f.PG())},
		} {
			val, src, err := db.DualScalar(ctx, "val", s.pg, f.Args()...)
			if err != nil {
				respondErr(w, 500, "Query failed: "+s.key)
				return
			}
			kpis[s.key] = val
			sources = append(sources, src)
		}

		// MTD always uses current month
		mtd, src, _ := db.DualScalar(ctx, "val",
			`SELECT COUNT(*) AS val FROM app.accounts WHERE DATE_TRUNC('month',opened_date)=DATE_TRUNC('month',CURRENT_DATE)`)
		kpis["issued_mtd"] = mtd
		sources = append(sources, src)

		total := toFloat(kpis["total_issued"])
		if total > 0 {
			kpis["activation_rate"] = round1(toFloat(kpis["active"]) / total * 100)
		} else {
			kpis["activation_rate"] = 0.0
		}

		respond(w, kpis, pickSource(sources))
	}
}

func cardTrendsIssuance(db *core.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		f, err := buildTrendsFilter(r)
		if err != nil {
			respondErr(w, 400, err.Error())
			return
		}
		data, src, err := db.DualQuery(r.Context(),
			fmt.Sprintf(`SELECT TO_CHAR(DATE_TRUNC('month',opened_date),'Mon YYYY') AS month,
			        DATE_TRUNC('month',opened_date) AS month_sort, COUNT(*) AS issued
			 FROM app.accounts WHERE opened_date IS NOT NULL%s
			 GROUP BY DATE_TRUNC('month',opened_date) ORDER BY month_sort`, f.PG()),
			f.Args()...)
		if err != nil {
			respondErr(w, 500, "Query failed")
			return
		}
		respond(w, data, src)
	}
}

func cardTrendsPortfolioHealth(db *core.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		f, err := buildTrendsFilter(r)
		if err != nil {
			respondErr(w, 400, err.Error())
			return
		}
		data, src, err := db.DualQuery(r.Context(),
			fmt.Sprintf(`SELECT TO_CHAR(DATE_TRUNC('month',opened_date),'Mon YYYY') AS month,
			        DATE_TRUNC('month',opened_date) AS month_sort,
			        SUM(CASE WHEN status IN ('Open','Active') THEN 1 ELSE 0 END) AS active,
			        SUM(CASE WHEN status NOT IN ('Open','Active') THEN 1 ELSE 0 END) AS inactive,
			        COUNT(*) AS total
			 FROM app.accounts WHERE opened_date IS NOT NULL%s
			 GROUP BY DATE_TRUNC('month',opened_date) ORDER BY month_sort`, f.PG()),
			f.Args()...)
		if err != nil {
			respondErr(w, 500, "Query failed")
			return
		}
		respond(w, data, src)
	}
}

func cardTrendsStatusDist(db *core.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		f, err := buildTrendsFilter(r)
		if err != nil {
			respondErr(w, 400, err.Error())
			return
		}
		data, src, err := db.DualQuery(r.Context(),
			fmt.Sprintf(`SELECT status AS status, COUNT(*) AS count FROM app.accounts WHERE 1=1%s GROUP BY status ORDER BY count DESC`, f.PG()),
			f.Args()...)
		if err != nil {
			respondErr(w, 500, "Query failed")
			return
		}
		respond(w, data, src)
	}
}

func cardTrendsByProgram(db *core.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		dateFrom, _ := validDate(r, "date_from")
		dateTo, _ := validDate(r, "date_to")
		var f Filter
		f.Date("Account_Created_Date", `opened_date`, dateFrom, dateTo)
		data, src, err := db.DualQuery(r.Context(),
			fmt.Sprintf(`SELECT COALESCE(card_product,card_program) AS program, COUNT(*) AS total,
			        SUM(CASE WHEN status IN ('Open','Active') THEN 1 ELSE 0 END) AS active,
			        SUM(CASE WHEN status NOT IN ('Open','Active') THEN 1 ELSE 0 END) AS inactive,
			        ROUND(100.0*SUM(CASE WHEN status IN ('Open','Active') THEN 1 ELSE 0 END)/COUNT(*),1) AS activation_rate
			 FROM app.accounts WHERE COALESCE(card_product,card_program) IS NOT NULL AND COALESCE(card_product,card_program)!=''%s
			 GROUP BY COALESCE(card_product,card_program) ORDER BY total DESC`, f.PG()),
			f.Args()...)
		if err != nil {
			data = []core.Row{} // non-fatal — column may not exist on all tenants
			src = "supabase_snapshot"
		}
		respond(w, data, src)
	}
}

func cardTrendsByProduct(db *core.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		f, err := buildTrendsFilter(r)
		if err != nil {
			respondErr(w, 400, err.Error())
			return
		}
		data, src, err := db.DualQuery(r.Context(),
			fmt.Sprintf(`SELECT product_name, COUNT(*) AS total,
			        SUM(CASE WHEN status IN ('Open','Active') THEN 1 ELSE 0 END) AS active,
			        SUM(CASE WHEN status NOT IN ('Open','Active') THEN 1 ELSE 0 END) AS inactive,
			        ROUND(100.0*SUM(CASE WHEN status IN ('Open','Active') THEN 1 ELSE 0 END)/COUNT(*),1) AS activation_rate
			 FROM app.accounts WHERE product_name IS NOT NULL%s
			 GROUP BY product_name ORDER BY total DESC`, f.PG()),
			f.Args()...)
		if err != nil {
			respondErr(w, 500, "Query failed")
			return
		}
		respond(w, data, src)
	}
}

func cardTrendsPrograms(db *core.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		data, src, err := db.DualQuery(r.Context(),
			`SELECT DISTINCT COALESCE(card_product,card_program) AS program FROM app.accounts
			 WHERE COALESCE(card_product,card_program) IS NOT NULL AND COALESCE(card_product,card_program)!='' ORDER BY COALESCE(card_product,card_program)`)
		if err != nil {
			respond(w, []string{}, "supabase_snapshot")
			return
		}
		programs := make([]any, 0, len(data))
		for _, row := range data {
			programs = append(programs, row["program"])
		}
		respond(w, programs, src)
	}
}
